server/src/test/java/org/elasticsearch/cluster/coordination/Zen1IT.java

```diff
@@ -45,4 +66,198 @@ public void testZen1NodesJoiningZen2Cluster() {
+            ensureStableCluster(nodeCount);
+            ensureGreen("test");
+            logger.info("--> successfully replaced {} with {}", zen1Node, newNode);
+        }
+
+        assertThat(internalCluster().size(), equalTo(nodeCount));
+    }
+
+    public void testMigratingFromZen1ToZen2ClusterWithTwoNodes() throws Exception {
+        testRollingMigrationFromZen1ToZen2(2);
+    }
+
+    public void testMigratingFromZen1ToZen2ClusterWithThreeNodes() throws Exception {
+        testRollingMigrationFromZen1ToZen2(3);
+    }
+
+    public void testMigratingFromZen1ToZen2ClusterWithFourNodes() throws Exception {
+        testRollingMigrationFromZen1ToZen2(4);
+    }
+
+    public void testMigratingFromZen1ToZen2ClusterWithFiveNodes() throws Exception {
+        testRollingMigrationFromZen1ToZen2(5);
+    }
+
+    private void testRollingUpgradeFromZen1ToZen2(final int nodeCount) throws Exception {
+        final List<String> nodes = internalCluster().startNodes(nodeCount, ZEN1_SETTINGS);
+
+        createIndex("test",
+            Settings.builder()
+                .put(UnassignedInfo.INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.getKey(), TimeValue.ZERO) // assign shards
+                .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, nodeCount) // causes rebalancing
+                .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 1)
+                .build());
+        ensureGreen("test");
+
+        internalCluster().rollingRestart(new RestartCallback() {
+            @Override
+            public void doAfterNodes(int n, Client client) {
+                ensureGreen("test");
+            }
+
+            @Override
+            public Settings onNodeStopped(String nodeName) {
+                String viaNode = randomValueOtherThan(nodeName, () -> randomFrom(nodes));
+                final ClusterHealthRequestBuilder clusterHealthRequestBuilder = client(viaNode).admin().cluster().prepareHealth()
+                    .setWaitForEvents(Priority.LANGUID)
+                    .setWaitForNodes(Integer.toString(nodeCount - 1))
+                    .setTimeout(TimeValue.timeValueSeconds(30));
+                if (nodeCount == 2) {
+                    clusterHealthRequestBuilder.setWaitForYellowStatus();
+                } else {
+                    clusterHealthRequestBuilder.setWaitForGreenStatus();
+                }
+                ClusterHealthResponse clusterHealthResponse = clusterHealthRequestBuilder.get();
+                assertFalse(nodeName, clusterHealthResponse.isTimedOut());
+                return Coordinator.addZen1Attribute(false, Settings.builder().put(ZEN2_SETTINGS)).build();
+            }
+        });
+
+        ensureStableCluster(nodeCount);
+        ensureGreen("test");
+        assertThat(internalCluster().size(), equalTo(nodeCount));
+    }
+
+    public void testUpgradingFromZen1ToZen2ClusterWithTwoNodes() throws Exception {
+        testRollingUpgradeFromZen1ToZen2(2);
```

Comment:
same here, should we randomize the number of nodes? Perhaps this test suite is taking a bit too long o.w..

Reply:
Yes, reduced the number of tests in 49a41d1344c.